File: snippets/js/sort_insertion_sort.js
// insertion sort - https://en.wikipedia.org/wiki/Insertion_sort
// grows an output list, sorting each new element into its proper place

function doSorting(inputList) {
  return insertionSort(inputList);
}

function insertionSort(inputList) {
  outputList = [];

  inputList.forEach((element, idx) => {
    outputList.push(element);
    outputList = sortElementAtIndex(outputList, element, idx);
  });

  return outputList;
}

function sortElementAtIndex(inputList, element, idx) {
  target_index = idx;
  outputList = inputList;

  while (target_index != 0 && element < outputList[target_index - 1]) {
    swapWithPrevious(outputList, target_index);
    target_index -= 1;
  }

  return outputList;
}

function swapWithPrevious(list, idx) {
  tmp = list[idx - 1];
  list[idx - 1] = list[idx];
  list[idx] = tmp;
  return list;
}

File: snippets/ruby/sort_insertion_sort.rb
# insertion sort - https://en.wikipedia.org/wiki/Insertion_sort
# steps through the input, inserting each element into its place in the output list

def do_sorting(input_list)
  insertion_sort(input_list)
end

def insertion_sort(input_list)
  output_list = []

  input_list.each_with_index do |element, idx|
    output_list.append(element)
    output_list = sort_element_at_index(output_list, element, idx)
  end

  output_list
end

def sort_element_at_index(input_list, element, idx)
  output_list = input_list
  target_index = idx

  while (target_index != 0) && (element < output_list[target_index - 1])
    output_list.swap(target_index)
    target_index -= 1
  end

  output_list
end

# add swap method to array
class Array
  def swap(idx)
    self[idx], self[idx - 1] = self[idx - 1], self[idx]
  end
end
